### PR TITLE
feat(ValidUser): also display badges & banner

### PR DESCRIPTION
**Pull request by dolfies**
### Summary
Fetching a user ignores the accent color, banner, and premium type given. We can also deduce nearly all badges from the flags (like it used to work :p, but means the badge text will always be English).

image

### CODE REVIEW

File: src/plugins/validUser/index.tsx [hunk 1]
```diff
@@ -21,12 +21,50 @@ import { Devs } from "@utils/constants";
...
+    "hypesquad_online_house_2": { id: "hypesquad_house_2", description: "HypeSquad Brilliance", icon: "011940fd013da3f7fb926e4a1cd2e618", link: "https://discord.com/settings/hypesquad-online" },
+    "hypesquad_online_house_3": { id: "hypesquad_house_3", description: "HypeSquad Balance", icon: "3aa41de486fa12454c3761e8e223442e", link: "https://discord.com/settings/hypesquad-online" },
+    "partner": { id: "partner", description: "Partnered Server Owner", icon: "3f9748e53446a137a052f3454e2de41e", link: "https://discord.com/partners" },
+    "premium": { id: "premium", description: "Subscriber", icon: "2ba85e8026a8614b640c2837bcdfe21b", link: "https://discord.com/settings/premium" },
+    "premium_early_supporter": { id: "early_supporter", description: "Early Supporter", icon: "7060786766c9c840eb3019e725d2b358", link: "https://discord.com/settings/premium" },
+    "verified_developer": { id: "verified_developer", description: "Early Verified Bot Developer", icon: "6df5892e0f35b051f8b61eace34f4967" },
+};
+
+enum UserFlags {
```
Comment (Vendicated) [inline src/plugins/validUser/index.tsx:43]: use a plain object

File: src/plugins/validUser/index.tsx [hunk 2]
```diff
@@ -43,6 +81,50 @@ interface MentionProps {
...
+    await FluxDispatcher.dispatch(
+        {
+            type: "USER_PROFILE_FETCH_FAILURE",
+            userId: id,
+        }
+    );
+
+    userObj = UserStore.getUser(id);
+    const fakeBadges: ProfileBadge[] = [];
+    for (const key in UserFlags) {
```
Comment (Vendicated) [inline src/plugins/validUser/index.tsx:108]:
```suggestion
    for (const [key, value] of Object.entries(UserFlags)) {
```

File: src/plugins/validUser/index.tsx [hunk 3]
```diff
@@ -85,14 +167,14 @@ function MentionWrapper({ data, UserMention, RoleMention, parse, props }: Mentio
...
-                            UserUtils.getUser(id)
+                            getUser(id)
                                 .then(() => {
                                     setUserId(id);
                                     fetching.delete(id);
                                 })
                                 .catch(e => {
                                     if (e?.status === 429) {
-                                        queue.unshift(() => sleep(1000).then(fetch));
+                                        queue.unshift(() => sleep(e?.body?.retry_after ?? 1000).then(fetch));
```
Comment (Vendicated) [inline src/plugins/validUser/index.tsx:177]:
```suggestion
                                        queue.unshift(() => sleep(Math.max(e?.body?.retry_after ?? 0, 1000)).then(fetch));
```

Comment (dolfies) [reply src/plugins/validUser/index.tsx:177]: What's wrong with waiting the rate limit? That's what they're for :p

**Review by Vendicated — COMMENTED** [review]
pretty cool!

File: src/plugins/validUser/index.tsx [hunk 4]
```diff
@@ -43,6 +81,47 @@ interface MentionProps {
     UserMention: ComponentType<any>;
 }
 
+async function getUser(id: string) {
+    let userObj = UserStore.getUser(id);
+    if (userObj)
+        return userObj;
+
+    const user: any = await RestAPI.get({ url: `/users/${id}` }).then(response => {
```
[the remark this answers is not in the document]
Comment (dolfies) [reply src/plugins/validUser/index.tsx:89]: yop

File: src/plugins/validUser/index.tsx [hunk 5]
```diff
@@ -43,6 +81,47 @@ interface MentionProps {
...
+            userId: id,
+        }
+    );
+
+    userObj = UserStore.getUser(id);
+    const fakeBadges: ProfileBadge[] = [];
+    for (const [key, value] of Object.entries(UserFlags)) {
+        if (userObj.hasFlag(value))
+            fakeBadges.push(badges[key.toLowerCase()]);
+    }
```
[the remark this answers is not in the document]
Comment (dolfies) [reply src/plugins/validUser/index.tsx]: done!

**Comment by Vendicated** [comment]
awesome, seems to work great! thank u mister alien 👽 💚

**Pull request merged**